src/spartan/mod.rs

```diff
@@ -84,25 +114,54 @@ impl<E: Engine> PolyEvalWitness<E> {
 }
 
 /// A type that holds a polynomial evaluation instance
-pub struct PolyEvalInstance<E: Engine> {
+struct PolyEvalInstance<E: Engine> {
   c: Commitment<E>,  // commitment to the polynomial
   x: Vec<E::Scalar>, // evaluation point
   e: E::Scalar,      // claimed evaluation
 }
 
 impl<E: Engine> PolyEvalInstance<E> {
-  fn pad(U: Vec<PolyEvalInstance<E>>) -> Vec<PolyEvalInstance<E>> {
-    // determine the maximum size
-    if let Some(ell) = U.iter().map(|u| u.x.len()).max() {
-      U.into_iter()
-        .map(|mut u| {
-          let mut x = vec![E::Scalar::ZERO; ell - u.x.len()];
-          x.append(&mut u.x);
-          PolyEvalInstance { x, ..u }
-        })
-        .collect()
-    } else {
-      Vec::new()
+  fn batch_diff_size(
+    c_vec: &[Commitment<E>],
+    e_vec: &[E::Scalar],
+    num_vars: &[usize],
+    x: Vec<E::Scalar>,
+    s: E::Scalar,
+  ) -> PolyEvalInstance<E> {
+    let num_instances = num_vars.len();
+    assert_eq!(c_vec.len(), num_instances);
+    assert_eq!(e_vec.len(), num_instances);
+
+    let num_vars_max = x.len();
+    let powers: Vec<E::Scalar> = powers::<E>(&s, num_instances);
+    // Rescale evaluations by the first Lagrange polynomial,
```

Comment:
I'm a bit confused about this rescaling. This wasn't necessary in the prior version (https://github.com/microsoft/Nova/blob/8244d00a7b3bf5fc44e8cf76c3a0d67cfdb17394/src/spartan/mod.rs#L109) didn't need this rescaling. Can you elaborate what changed?

Reply:
During batched Sumcheck over `n` variables, we a interpret polynomial `P` with `m<n` as having `n-m` variables inserted "in front" (the original `m` variables are `X_{n-m}, ..., X_{n-1}`). This means that Sumcheck will return the evaluation `P(r_{n-m}, ..., r_{n-1})`. However, we are using a PCS that supports polynomials in `n` variables, and we committed to `P` by doing an MSM with the first `2^m` entries of the commitment key. This means we really committed to the polynomial `P'(X_0, ..., X_{n-1}) = L_0(X_0, ..., X_{n-m-1}) * P(X_{n-m}, ..., X_{n-1})`, where only the first `2^m` entries are non-zero. 

The degree bounding checks for `P` are all implicitly performed by the Sumcheck instances, since we use evaluation of `eq` with the appropriate number of variables, all computed by the verifier. The scaling that happens in `batch` is just a way of allowing for the two interpretations `P` and `P'`.

Reply:
Hi Adrian! Thanks for the response! I understood the scaling part (in a way, this is similar to the scaling of IO-related polynomials when combining W and IO to account for padding, right?).

What is still confusing me is that we didn't need to do such scaling in the prior version (linked above). Can you elaborate why we need scaling of evaluations here but we didn't need it before?

Reply:
That's right, the idea is essentially the same as the `W` and `IO` vectors. 

In the prior version, all vectors were padded with zeros to the same `N = 2^n`. A polynomial `P` that only had `m < n` variables would be interpreted from the start as `L_0(X_0, ..., X_{n-m-1})P(X_{n-m}, ..., X_{n-1})`, including during Sumcheck. Therefore, this scaling is applied by Sumcheck during the `bind` phase. In the first `n-m` rounds, the second half of the polynomial will be 0, so the effect is just a multiplication of the vector by `(1-r_i)`. 

The batched Sumcheck handles the unpadded `P` and only evaluates it in `r_{n-m}, ..., r_{n-1}`, avoiding the multiplication by `(1-r_i)` in the first `n-m` rounds. 

Essentially, padding with zeros will force Sumcheck to do the scaling. Since we avoid the padding, we need to manually add the scaling by the Lagrange polynomials in the batch open phase.